### ISSUE TITLE
Proposal: Habitat post run hook

### ISSUE THREAD

**Issue by dmajere** [habitat-sh/habitat#1393]
Sometimes we need to run not only some tasks before application will run, but also after.
It's would be nice to have postinit hook for habitat.
So the flow would be
init -> run -> post

**Comment by bdangit** [comment]
what could those tasks be?

**Comment by dmajere** [comment]
Well, I have it mostly with erlang applications, when you need to start erlang node first and then do something with it. A good example, probably, is rabbitmq, you need to have it started before you'll be able to do rabbitctl create_vhost and so on.

**Comment by bdangit** [comment]
fair point! I've seen some plans where people will do this during `init` hook where they would start the service under dameonize mode, configure it, shut it down and then let `run` run the service under `hab-sup`.  Its clunky but it does qualify as "init".  

The `post` might be tricky unless `hab-sup` is able to spawn another task to do what you want.

**Comment by eeyun** [comment]
I'm very curious about this issue. I'd like to get some more community weigh in on it. It would be good to discuss whether or not its a realistic thing. It _might_ be tricky. Unfortunately the supervisor is the piece of the codebase I have the LEAST exposure to and is simultaneously the most in flux currently :).

**Comment by reset** [comment]
@dmajere we've been adding lifecycle hooks as we needed them but haven't found a true use for `after-run` just yet. The problem that I see with adding any lifecycle hooks which happen after `run` is that you lose the ability to coordinate in case of a failure. Basically anything that would happen in `after-run` can't be critical to the service.

Initializing a database's schema is actually not the job of the database service itself, but the consumer of the database service. We actually added a topology for just this called the "initializer topology". Also as @bdangit points out, the init hook can be leveraged in clever ways to ensure all initialization steps are taken before the service runs.

I'm all for adding additional lifecycle hooks - let's just make sure we've got some concrete examples of where to use them. If we need to modify the `init` hook into multiple steps to add clarity we could investigate that as well.
